### ISSUE TITLE
Resolve SimplePaths to DefId's

### ISSUE THREAD

**Issue by philberty** [Rust-GCC/gccrs#1159]
This change depends on #1158 

- [x] Add missing mappings to HIR::SimplePath #1176 
- [x] Add missing mappings to HIR::SimpePathSegment #1176
- [x] Remove the default locus argument of builtin_location on HIR::SimplePathSegment #1174 

Resolving a SimplePath to a DefId its achieved via the name-resolver class:

```
   Parameter: SimplePath* expr;

    NodeId ast_node_id = expr.get_mappings ().get_nodeid ();

    // then lookup the reference_node_id
    NodeId ref_node_id = UNKNOWN_NODEID;
    if (resolver->lookup_resolved_name (ast_node_id, &ref_node_id))
      {
	// these ref_node_ids will resolve to a pattern declaration but we are
	// interested in the definition that this refers to get the parent id
	Definition def;
	if (!resolver->lookup_definition (ref_node_id, &def))
	  {
	    // FIXME
	    // this is an internal error
	    rust_error_at (expr.get_locus (),
			   "unknown reference for resolved name");
	    return;
	  }
	ref_node_id = def.parent;
      }
```

if we are assuming this path can or should only resolve to modules we can ignore the type resolver name-space.

```
    if (ref_node_id == UNKNOWN_NODEID)
      {
        // this is an internal error but error messages are useful for now
        // down the line we will be able to rely on the changes make in #1158 
        // that this will have already errored out for us and we can just return
	rust_error_at (expr.get_locus (), "path: %s",
		       expr.as_string ().c_str ());
	return;
      }
```

Now we have resolved the expression to its definitions NodeId which is all within the AST. 

### Getting the DefId from NodeId

Note there is a bug with our mappings class here but we can fix this either as part of this issue or in a separate issue/pr we need to reverse map the NodeId back to an Hir node so that we can look up the Hir node to the DefID but the mappings class as nested these reverse mappings with the associated CrateNum which is wrong but we can ignore this for now.

```
    // node back to HIR
    HirId ref;
    if (!mappings->lookup_node_to_hir (expr.get_mappings ().get_crate_num (),
				       ref_node_id, &ref))
      {
	// this is an internal error like above but its useful for debugging for now
	rust_error_at (expr.get_locus (), "reverse lookup failure");
	return;
      }
```

Now that we have the associated HirId of the resolved path we can look it up again in the mappings and again this wrongly associates the crate num here but ignore this for now.

```
  HIR::Item *lookup_hir_item (CrateNum crateNum, HirId id);
```
Now you have an HIR::Item and you can get the DefId directly from that item which should be the module you looking for.

**Comment by CohenArthur** [comment]
The three base issues of this are done, but we still need to do the actual resolving :) Hence why I ticked all boxes but left the issue open
